File: atividade-21/atividade.go
package main

import "fmt"

type AFN struct {
	states       map[string]bool
	startState   string
	acceptStates map[string]bool
}

func NewAFN() *AFN {
	return &AFN{
		states: map[string]bool{
			"q0":       true,
			"q1":       true,
			"q_reject": true,
		},
		startState: "q0",
		acceptStates: map[string]bool{
			"q0": true,
		},
	}
}

func (afn *AFN) isAccepted(input string) bool {
	currentStates := map[string]bool{
		afn.startState: true,
	}

	for _, c := range input {
		nextStates := make(map[string]bool)
		for state := range currentStates {
			if state == "q0" {
				if c == 'a' {
					nextStates["q0"] = true
				} else if c == 'b' {
					nextStates["q1"] = true
				}
			} else if state == "q1" {
				if c == 'a' {
					nextStates["q0"] = true
				} else if c == 'b' {
					nextStates["q_reject"] = true
				}
			} else if state == "q_reject" {
				nextStates["q_reject"] = true
			}
		}
		currentStates = nextStates
	}

	return currentStates["q0"]
}

func main() {
	afn := NewAFN()
	fmt.Println(afn.isAccepted("a"))     
	fmt.Println(afn.isAccepted("ab"))    
	fmt.Println(afn.isAccepted("abb"))   
	fmt.Println(afn.isAccepted("ababa")) 
	fmt.Println(afn.isAccepted("bba"))   
	fmt.Println(afn.isAccepted("ba"))    
	fmt.Println(afn.isAccepted("b"))     
}

File: atividade-21/atividade.py
class AFN:
    def __init__(self):
        self.states = {'q0', 'q1', 'q_reject'}
        self.start_state = 'q0'
        self.accept_states = {'q0'}
    
    def is_accepted(self, string):
        current_states = {self.start_state}

        for char in string:
            next_states = set()
            for state in current_states:
                if state == 'q0':
                    if char == 'a':
                        next_states.add('q0')
                    elif char == 'b':
                        next_states.add('q1')
                elif state == 'q1':
                    if char == 'a':
                        next_states.add('q0')
                    elif char == 'b':
                        next_states.add('q_reject')
                elif state == 'q_reject':
                    next_states.add('q_reject')
            current_states = next_states

        return 'q0' in current_states

if __name__ == '__main__':
    afn = AFN()
    print(afn.is_accepted("a"))       
    print(afn.is_accepted("ab"))      
    print(afn.is_accepted("abb"))     
    print(afn.is_accepted("ababa"))    
    print(afn.is_accepted("bba"))    
    print(afn.is_accepted("ba"))      
    print(afn.is_accepted("b"))       